torch/testing/_internal/common_cuda.py

```diff
@@ -32,3 +35,43 @@ def initialize_cuda_context_rng():
         for i in range(torch.cuda.device_count()):
             torch.randn(1, device="cuda:{}".format(i))
         __cuda_ctx_rng_initialized = True
+
+
+def tf32_is_not_fp32():
+    if not torch.cuda.is_available():
+        return False
+    if torch.cuda.get_device_properties(torch.cuda.current_device()).major < 8:
+        return False
+    return True
+
+
+@contextlib.contextmanager
+def setup_tf32(dtype, rtol=0.001, atol=1e-5):
+    old = torch.backends.cuda.matmul.allow_tf32
+    if dtype in {tfloat32, tcomplex64}:
+        torch.backends.cuda.matmul.allow_tf32 = True
+        if tf32_is_not_fp32():
+            yield tf32_to_fp32(dtype), rtol, atol
+        else:
+            yield tf32_to_fp32(dtype), None, None
+    else:
+        if dtype in {torch.float32, torch.complex64}:
+            torch.backends.cuda.matmul.allow_tf32 = False
+        yield dtype, None, None
+    torch.backends.cuda.matmul.allow_tf32 = old
+
+
+def tf32_on_and_off(rtol_=0.001, atol_=1e-5):
+    def wrapper(f):
+        @functools.wraps(f)
+        def wrapped(*args, **kwargs):
+            old = torch.backends.cuda.matmul.allow_tf32
+            rtol, atol = (rtol_, atol_) if tf32_is_not_fp32() else (None, None)
+            torch.backends.cuda.matmul.allow_tf32 = True
+            f(*args, **kwargs, rtol=rtol, atol=atol)
```

Comment:
So this decorator will run any test that's decorated with it twice, regardless of whether it is needed or not? (e.g. cpu tests would run twice, tests on non-ampere hardware would run twice). Can you make tf32 a dtype?

Reply:
You should check here that:

- no dtype is passed to the test or the passed dtype is fp32
- you have ampere hardware
- the compiled CUDA supports TF32

Also, you should either:

- implement a context manager for allow_tf32
- OR reset it in a finally block

That will prevent the global state from leaking if the test fails with a runtime error.

Reply:
if you go with a `contextlib.contextmanager`, be careful of https://github.com/pytorch/pytorch/pull/40800#discussion_r448607374.

Reply:
fixed